File: backend/app/schemas/report.py
from datetime import datetime

from pydantic import BaseModel, Field


class DatasetSummary(BaseModel):
    id: str
    name: str
    row_count: int
    columns: list[str]
    sample_rows: list[dict] = Field(default_factory=list)


class ToolEvent(BaseModel):
    tool: str
    detail: str


class ReportSection(BaseModel):
    id: str
    title: str
    markdown: str


class ReportChart(BaseModel):
    id: str
    title: str
    chart_type: str
    spec: dict
    image_data_url: str | None = None
    query_id: str | None = None


class CreateReportRequest(BaseModel):
    prompt: str
    datasets: list[DatasetSummary]


class CreateReportResponse(BaseModel):
    id: str
    status: str
    sections: list[ReportSection]
    charts: list[ReportChart]
    tool_log: list[ToolEvent]


class ReportResponse(CreateReportResponse):
    prompt: str
    uploaded_files: list[dict]
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/report.ts
import type { ClientChartSpec, DataRow } from "./analysis";

export type LocalAttachmentKind = "csv" | "json" | "pdf" | "image" | "other";
export type DatasetKind = "csv" | "json";

export type LocalAttachmentSummary = {
  id: string;
  name: string;
  kind: LocalAttachmentKind;
  extension: string;
  byte_size?: number;
  mime_type?: string;
};

export type DatasetSummary = LocalAttachmentSummary & {
  kind: DatasetKind;
  row_count: number;
  columns: string[];
  numeric_columns: string[];
  sample_rows: DataRow[];
};

export type LocalDataset = DatasetSummary & {
  rows: DataRow[];
  preview_rows: DataRow[];
  json_text?: string;
};

export type LocalPdfAttachment = LocalAttachmentSummary & {
  kind: "pdf";
  page_count: number;
  bytes_base64: string;
};

export type LocalImageAttachment = LocalAttachmentSummary & {
  kind: "image";
  width: number;
  height: number;
  bytes_base64: string;
};

export type LocalOtherAttachment = LocalAttachmentSummary & {
  kind: "other";
  text_content?: string;
  bytes_base64?: string;
};

export type LocalAttachment =
  | LocalDataset
  | LocalPdfAttachment
  | LocalImageAttachment
  | LocalOtherAttachment;

export type ReportSection = {
  id: string;
  title: string;
  markdown: string;
};

export type ReportChart = {
  id: string;
  title: string;
  chart_type: string;
  spec: ClientChartSpec | Record<string, unknown>;
  image_data_url?: string | null;
  dataset_id?: string | null;
  chart_plan_id?: string | null;
};

export type ToolEvent = {
  tool: string;
  detail: string;
};
